libbpf-tools/funcslower.c

```diff
@@ -0,0 +1,624 @@
+        }
+        switch (env.units) {
+        case NSEC:
+                duration = event->duration_ns / 1.0;
+                break;
+        case USEC:
+                duration = event->duration_ns / 1000.0;
+                break;
+        case MSEC:
+                duration = event->duration_ns / 1000000.0;
+                break;
+        default:
+                warn("bad units");
+        };
+
+        printf("%-14.14s %-8d %7.2f %8llx %s", event->comm,
+               (int)((event->tgid) >> 32), duration, event->retval,
+               env.funcnames[event->id]);
+        for (int i = 0; i < env.args; ++i) {
+                printf(" %llx ", event->args[i]);
+        }
+        printf("\n");
+
+        if (env.kstack || env.ustack) {
+                if (!env.ip) {
+                        env.ip = calloc(env.perf_max_stack_depth, sizeof(unsigned long));
+                        if (!env.ip) {
+                                fprintf(stderr, "failed to alloc ip\n");
+                                return;
+                        }
+                }
+                if (env.kstack) {
+                        if (bpf_map_lookup_elem(env.stack_map_fd, &event->kstack, env.ip) != 0) {
+                                fprintf(stderr, "    [Missed Kernel Stack]\n");
+                        }
+                        print_kstack(env.ksyms, env.ip, env.perf_max_stack_depth, 0);
+                }
+                if (env.ustack) {
+                        if (bpf_map_lookup_elem(env.stack_map_fd, &event->ustack, env.ip) != 0) {
+                                fprintf(stderr, "    [Missed User Stack]\n");
+                        }
+                        print_ustack(env.syms_cache, event->tgid, env.ip, env.perf_max_stack_depth, 0);
+                }
+        }
+}
+
+int main(int argc, char **argv) {
+
+        LIBBPF_OPTS(bpf_object_open_opts, open_opts);
+        int err;
+        int idx, cg_map_fd;
+        int cgfd = -1;
+        struct funcslower_bpf **obj;
+        struct perf_buffer *pb = NULL;
+        int key = 0;
+        int inner_map_event_fd, inner_map_stack_fd;
+        struct bpf_map *outer_stack_map, *outer_event_map;
+        int outer_stack_map_fd, outer_event_map_fd;
+        const char *outer_stack_map_name = "outer_stack_map";
+        const char *outer_event_map_name = "outer_event_map";
+        static const struct argp argp = {
+            .options = opts,
+            .parser = parse_arg,
+            .args_doc = args_doc,
+            .doc = program_doc,
+        };
+
+        err = argp_parse(&argp, argc, argv, 0, NULL, &env);
+        if (err)
+                return err;
+
+        sigaction(SIGINT, &sigact, 0);
+
+        libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+        libbpf_set_print(libbpf_print_fn);
+
+        obj = calloc(env.num_funcs, sizeof(struct funcslower_bpf *));
+        if (!obj) {
+                fprintf(stderr, "failed to create obj array\n");
+                return 1;
+        }
```

Comment:
No, you can attach a BPF program to multiple target.

Reply:
yes, of course. but problem here is there is no clear way to differentiate different functions we trace. here it's my proposed way of addressing this. maybe a better way?

Reply:
actually i come up with another way to handle the identification problem via bss variable, which should be conciser. will push the changes later.

Reply:
i think i've found a better way to do this with existed in-kernel support of bpf cookies. will push new version later.